### DISCUSSION TITLE
Does using dotNET 8 (or even 6 or 5) stops me from publishing to consoles and mobile?

### DISCUSSION THREAD

**Discussion by roo-oliv** [MonoGame/MonoGame#8265] [General]
If I’m not mistaken, nothing is preventing one from using MonoGame 3.8 targeting **dotnet 8** but I’m confused if the dotnet version will stop me from releasing my games on other platforms other than desktop (Mac, Windows, Linux) such as **consoles** (Switch, Playstation, Xbox) and **mobile** (Android and iOS).

I’d like to use the latest dotnet but only if this will not be an impedance to port my games. So, **can I use whatever dotnet version I want** or is it better to stick to a specific version such as dotnet 4, 5, or 6?

**Comment by ThomasFOG** [comment]
It really depends on your schedule.

**For mobile, short and straight answer**

You _should_ use .NET 8.

Working with previous versions is deprecated and likely to be refused by Google Play and the Apple Store because it can't target the required SDK versions.

**Regarding consoles**

No matter the timeline, there's one thing to be aware of: **you won't be able to use runtime reflection there** (to verify, your game must build _and_ run without crashes with ```PublishAot``` and all warnings of ```EnableTrimAnalyzer``` fixed).

That said, there's two scenario which depend on how urgent releasing your game is:

- If you need to port to consoles _now_: using .NET 8 is fine but you will have to use only C# 5.0 syntax (you can force it using ```<LangVersion>5</LangVersion>```). You also can't use namespaces or classes that aren't present in .NET 4.5.2 (you can force targeting ```net452``` to verify that you're not).

- If you can afford waiting a few months: .NET 8 is recommended and not limited to C# 5.0 or namespaces etc.

**Reply by roo-oliv** [reply]
Thanks for the quick and clear reply @mrhelmut! Also, I admire your dedication and work in MonoGame, congrats :heart: 

I sure can wait since I'm releasing to Desktop later this year and consoles in 2025.

So just to confirm: I can use .NET 8 and C# 12, I don't need to worry about any caveats such as specific .NET classes and namespaces, and I can use any valid C# 12 syntax (or there is just a new ceiling, like C# 9?). All I have to do [if I can wait] is just making sure that everything works when NativeAOT compilation is enabled and `EnableTrimAnalyzer`  isn't issuing any warnings. Which basically is staying clear of any use of reflection.

Is that right?

**Reply by Brett208** [reply]
@roo-oliv, 

We were recently released a MonoGame project on Apple, Google Play, and Amazon using a current MonoGame release, .Net 8, C#12 and AOT compilation, so yes this is absolutely doable. We were able to use the same codebase for a UWP release on Xbox as well (there is nuance in what it means to use .net 8 and C#12 and be compatible with UWP.) I think UWP will be deprecated within MonoGame in the near future so keep that in mind.

For AOT related activity, you will have to be careful to set some trimmer roots, below is what we set for iOS. I think there is work to improve trimming ongoing but it isn't in the baseline of MonoGame yet.
```
  <ItemGroup>
    <TrimmerRootAssembly Include="MonoGame.Framework" />
    <TrimmerRootAssembly Include="mscorlib" />
  </ItemGroup>
```

I cannot speak to console development as we didn't go that route besides UWP.

While MonoGame will do a lot of heavy lifting enabling cross-platform compatibility, there will always be some discrepancies in how the platforms operate and one interfaces with them. If you are planning to release on such a wide variety of platforms, I would encourage setting up automated builds using GitHub Runners or equivalent technology very early in the project life. Also complete some early on device testing. This will catch any of the anomalies between the platforms in your code base early and allow for easier workarounds.

Unfortunately, writing the automated build scripts is not trivial if you haven't worked on this before (at least for me). You would need the NDAs in place to attempt on the consoles. But even without the consoles up front at least you could gain confidence in the iOS, Android, and Windows builds.

Hope this helps.
~Brett

**Reply by roo-oliv** [reply]
Thanks for the thorough answer @Brett208 ! 😃 I planned consoles for 2025 anticipating the time needed to get the NDAs signed and hands on the consoles. But I'll try to get them faster so I can test my builds early.

Your experience is very encouraging! I don't plan to go for UWP so I'll start testing net8.0 and C#12 on Linux, MacOS, Windows, Android, and iOS for now!

Thank you all, this was very enlightening! 💡

**Reply by ThomasFOG** [reply]
> Is that right?

Correct!

Though mind that NativeAOT is "community-ported" on consoles and that it's not a 100% coverage and accuracy compared to its PC version, so there are chances that some features are in blind spots. It should work, just don't expect it to be magical.
